File: Python/groom/1차대회-startup-coding-festival/6.py
N, M = map(int, input().split()) # 가로 N, 세로 M

clothes = [[*map(int, input().split())] for _ in range(M)]

cache = dict()

def solve(y, x):
    if not (0 <= y < M and 0 <= x < N):
        return 0

    if (y, x) not in cache:
        res = max(solve(y-1, x), solve(y, x-1))
        cache[(y, x)] = res + clothes[y][x]

    return cache[(y, x)]

# print(solve(M-1, N-1))

clothes = [[1] * 100] * 10000

N = 100
M = 10000

print(solve(M-1, N-1))

File: CPP/groom/first_contest/6.cpp
#include <iostream>
#include <vector>
#include <algorithm>

typedef long long ll;
ll cache[10000][100] = { -1, };
int clothes[10000][100] = { -1, };

using namespace std;
int N, M;

int solve(int y, int x) {
    if (!(0 <= y && y < M && 0 <= x && x < N)) return 0;

    if (cache[y][x] == -1) {
        int result = max(solve(y-1, x), solve(y, x-1));
        cache[y][x] = result + clothes[y][x];
    }

    return cache[y][x];
}

int main() {
    cin >> N >> M;

    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            int val; cin >> val;
            cache[i][j] = -1;
            clothes[i][j] = val;
        }
    }

    cout << solve(M-1, N-1) << "\n";

    return 0;
}